File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.js
class TrieNode {
    constructor() {
        this.children = {};
        this.endOfWord = false;
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }


/** 
 * @param {string} word
 * @return {void}
 */
    insert(word) {
        let curr = this.root;
        
        // go through each char, make new node if needed
        for (const c of word) {
            if (!curr.children[c]) {
                curr.children[c] = new TrieNode();
            }
            curr = curr.children[c];
        }
        // curr is now last letter in word
        curr.endOfWord = true;
    };

/** 
 * @param {string} word
 * @return {boolean}
 */
    search(word) {
        let curr = this.root;
        
        // go through each char, 
        // return false instead of making new nodes
        for (const c of word) {
            if (!curr.children[c]) {
                return false;
            }
            curr = curr.children[c];
        }
        
        // curr should be last in word;
        
        return curr.endOfWord;
    };

/** 
 * @param {string} prefix
 * @return {boolean}
 */
    // this is why you need a prefix tree
    // allows for this func to be v efficient
    startsWith(prefix) {
        let curr = this.root;
        
        // go through each char like above
        for (const c of prefix) {
            if (!(c in curr.children)) {
                return false;
            }
            curr = curr.children[c];
        }
        
        // don't need to check end of word
        return true;
    };
};
/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.rb
class TrieNode
   attr_accessor :children, :is_word 
    
   def initialize()
       @children = {}
       @is_word = false
   end
end

class Trie
    def initialize()
        @root = TrieNode.new
    end


=begin
    :type word: String
    :rtype: Void
=end
    def insert(word)
        curr = @root
        word.each_char do |c|
            curr.children[c] = TrieNode.new if !curr.children[c]
            curr = curr.children[c]
        end
        curr.is_word = true
    end


=begin
    :type word: String
    :rtype: Boolean
=end
    def search(word)
        curr = @root
        word.each_char do |c|
            return false if !curr.children[c]
            curr = curr.children[c]
        end
        curr.is_word
    end


=begin
    :type prefix: String
    :rtype: Boolean
=end
    def starts_with(prefix)
        curr = @root
        prefix.each_char do |c|
            return false if !curr.children[c]
            curr = curr.children[c]
        end
        true
    end


end

# Your Trie object will be instantiated and called as such:
# obj = Trie.new()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.starts_with(prefix)